File: MIDTERM-LAB-3/MP14/MP14_KeywordCount.js
// ============================================================
// MP14 - Count Keyword Occurrences
// Programmer: OBNIAL, MARIUS RYZER C.
// Course: Programming 2 | BSIT GD - 1st Year
// Description: This program asks the user for a keyword,
//              then counts how many times it appears across
//              all rows and columns in the CSV dataset.
// ============================================================

// 'fs' = Node.js module for file operations (reading files)
const fs = require("fs");

// 'readline' = Node.js module for reading keyboard input
const readline = require("readline");

// Create the input/output interface for the terminal
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Step 1: Ask for the file path first
rl.question("Enter the CSV dataset file path: ", function (filePath) {

  // Step 2: Ask for the keyword to count
  rl.question("Enter keyword to count: ", function (keyword) {

    keyword = keyword.trim(); // Remove extra spaces from keyword

    // If nothing was typed, show error and stop
    if (keyword === "") {
      console.log("ERROR: No keyword entered.");
      rl.close();
      return;
    }

    // Call our main processing function
    processKeywordCount(filePath, keyword);

    rl.close(); // Close input interface after we're done asking
  });
});

// ============================================================
// FUNCTION: processKeywordCount
// PURPOSE : Reads the CSV file, counts how many times the
//           keyword appears, and displays the results.
// PARAMETERS:
//   filePath - path to the CSV file
//   keyword  - the word/phrase to search for
// ============================================================
function processKeywordCount(filePath, keyword) {

  // Check if the file actually exists at the given path
  if (!fs.existsSync(filePath)) {
    console.log("ERROR: File not found. Please check the file path.");
    return;
  }

  // Read the entire file as text
  const fileContent = fs.readFileSync(filePath, "utf8");

  // Split the text into an array of lines
  const lines = fileContent.split("\n");

  // Convert keyword to lowercase for case-insensitive search
  // e.g., searching "pass" will also match "PASS", "Pass", etc.
  const keywordLower = keyword.toLowerCase();

  // --- Counters and storage ---
  let totalOccurrences = 0;  // Grand total of all keyword matches
  let rowsContaining = 0;    // How many rows had at least one match
  let dataRowCount = 0;      // How many real data rows were checked
  let matchDetails = [];     // Stores info about each matched row
  let headerFound = false;   // Tracks if we've passed the header row

  // --- Process each line ---
  for (let i = 0; i < lines.length; i++) {
    let lineNumber = i + 1;        // Line number for display (starts at 1)
    let line = lines[i].trim();    // Current line, trimmed of whitespace

    // Skip blank lines
    if (line === "") continue;

    // Parse the line into individual column values
    let columns = parseCSVLine(line);

    // --- Find the real header row ---
    if (!headerFound && columns[0].trim().toLowerCase() === "candidate") {
      headerFound = true; // Mark that we found the header
      continue;           // Skip this line (it's a header, not data)
    }

    // --- Process data rows only (after header is found) ---
    if (headerFound) {
      dataRowCount++; // Count this row

      let countInThisRow = 0; // Matches found in just this row

      // Check each column value in the row
      for (let j = 0; j < columns.length; j++) {
        let cellLower = columns[j].trim().toLowerCase(); // Lowercase for comparison

        // Count how many times keyword appears in this cell
        let found = countOccurrences(cellLower, keywordLower);
        countInThisRow += found; // Add to this row's count
      }

      // If keyword appeared at least once in this row
      if (countInThisRow > 0) {
        totalOccurrences += countInThisRow; // Add to grand total
        rowsContaining++;                   // Count this row as a match

        let name = columns.length > 0 ? columns[0].trim() : "(unknown)";
        matchDetails.push(
          `Row ${lineNumber} | ${name} | Occurrences in row: ${countInThisRow}`
        );
      }
    }
  }

  // --- Display the results ---
  console.log("\n============================================================");
  console.log("           MP14 - KEYWORD OCCURRENCE COUNT");
  console.log("============================================================");
  console.log(`  Keyword Searched  : "${keyword}"`);
  console.log("  File              : " + filePath);
  console.log("  Total Data Rows   : " + dataRowCount);
  console.log("------------------------------------------------------------");
  console.log("  Total Occurrences : " + totalOccurrences);
  console.log("  Rows Containing   : " + rowsContaining);
  console.log("------------------------------------------------------------");

  if (matchDetails.length === 0) {
    console.log(`  No matches found for "${keyword}".`);
  } else {
    console.log("  Matched Rows:");
    matchDetails.forEach(function (detail) {
      console.log("    " + detail);
    });
  }

  console.log("============================================================");
}

// ============================================================
// FUNCTION: countOccurrences
// PURPOSE : Counts how many times 'keyword' appears in 'text'
// PARAMETERS:
//   text    - the string to search inside
//   keyword - the string to look for
// RETURNS: number (count of matches)
// ============================================================
function countOccurrences(text, keyword) {
  let count = 0;   // Start counting at zero
  let index = 0;   // Start searching from position 0 in the text

  // indexOf returns -1 when the keyword is not found anymore
  while ((index = text.indexOf(keyword, index)) !== -1) {
    count++;                   // Found a match!
    index += keyword.length;   // Move forward to search for the next one
  }

  return count; // Return total count
}

// ============================================================
// FUNCTION: parseCSVLine
// PURPOSE : Splits a CSV line into column values correctly,
//           handling quoted fields that may contain commas.
// PARAMETER: line - a single line from the CSV file
// RETURNS: array of string values
// ============================================================
function parseCSVLine(line) {
  let result = [];      // Stores all extracted values
  let current = "";     // Builds up the current field value
  let inQuotes = false; // Are we inside a quoted field?

  for (let i = 0; i < line.length; i++) {
    let char = line[i];

    if (char === '"') {
      inQuotes = !inQuotes; // Flip the flag when we see a quote
    } else if (char === "," && !inQuotes) {
      result.push(current); // End of field — save it
      current = "";          // Reset for next field
    } else {
      current += char; // Add character to current field
    }
  }

  result.push(current); // Save the last field
  return result;
}

File: MIDTERM-LAB-3/MP14/MP14_KeywordCount.java
// ============================================================
// MP14 - Count Keyword Occurrences
// Programmer: OBNIAL, MARIUS RYZER C.
// Course: Programming 2 | BSIT GD - 1st Year
// Description: This program asks the user for a keyword,
//              then searches the entire CSV dataset for how
//              many times that keyword appears across all rows
//              and columns.
// ============================================================

import java.io.*;        // For reading files
import java.util.*;      // For ArrayList, Scanner

public class MP14_KeywordCount {

    public static void main(String[] args) {

        // Scanner reads keyboard input from the user
        Scanner scanner = new Scanner(System.in);

        // Step 1: Ask for the file path
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine(); // Store the path user typed

        // Step 2: Ask for the keyword to search
        System.out.print("Enter keyword to count: ");
        String keyword = scanner.nextLine().trim(); // Store and clean the keyword

        // If the user typed nothing, exit early
        if (keyword.isEmpty()) {
            System.out.println("ERROR: No keyword entered.");
            scanner.close();
            return;
        }

        // 'keywordLower' is the keyword in lowercase so we can do
        // case-insensitive comparison (e.g., "pass" matches "PASS", "Pass")
        String keywordLower = keyword.toLowerCase();

        // --- Counters ---
        int totalOccurrences = 0;  // How many total times keyword was found
        int rowsContaining = 0;    // How many rows had at least one match
        int dataRowCount = 0;      // How many real data rows were processed

        // This list stores descriptions of each row that had a match
        ArrayList<String> matchDetails = new ArrayList<>();

        // 'headerFound' tells us if we've already passed the header line
        boolean headerFound = false;

        // --- Start reading the file ---
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line;       // Holds the current line
            int lineNumber = 0; // Tracks line number for display

            while ((line = br.readLine()) != null) {
                lineNumber++; // Count every line we read

                // Skip blank lines
                if (line.trim().isEmpty()) continue;

                // Split the line into column values
                // We use a simple split here; quoted commas are rare in this dataset
                String[] columns = line.split(",", -1);

                // --- Look for the header row ---
                if (!headerFound && columns[0].trim().equalsIgnoreCase("Candidate")) {
                    headerFound = true; // Found header, now look at data rows
                    continue;
                }

                // --- Only process real data rows (after the header) ---
                if (headerFound) {
                    dataRowCount++; // Count this as a data row

                    int countInThisRow = 0; // Keyword count just for this row

                    // Loop through each column value in this row
                    for (String col : columns) {
                        // Convert the cell value to lowercase for comparison
                        String cellLower = col.trim().toLowerCase();

                        // Count how many times the keyword appears in this cell
                        // We use countOccurrences() helper method below
                        int found = countOccurrences(cellLower, keywordLower);
                        countInThisRow += found; // Add to this row's count
                    }

                    // If we found the keyword at least once in this row
                    if (countInThisRow > 0) {
                        totalOccurrences += countInThisRow; // Add to grand total
                        rowsContaining++;                   // Mark this row as matched

                        // Get the candidate's name for display (first column)
                        String name = columns.length > 0 ? columns[0].trim() : "(unknown)";
                        matchDetails.add("Row " + lineNumber +
                                         " | " + name +
                                         " | Occurrences in row: " + countInThisRow);
                    }
                }
            }

            // --- Display the results ---
            System.out.println("\n============================================================");
            System.out.println("           MP14 - KEYWORD OCCURRENCE COUNT");
            System.out.println("============================================================");
            System.out.printf("  Keyword Searched  : \"%s\"%n", keyword);
            System.out.println("  File              : " + filePath);
            System.out.println("  Total Data Rows   : " + dataRowCount);
            System.out.println("------------------------------------------------------------");
            System.out.println("  Total Occurrences : " + totalOccurrences);
            System.out.println("  Rows Containing   : " + rowsContaining);
            System.out.println("------------------------------------------------------------");

            if (matchDetails.isEmpty()) {
                System.out.println("  No matches found for \"" + keyword + "\".");
            } else {
                System.out.println("  Matched Rows:");
                for (String detail : matchDetails) {
                    System.out.println("    " + detail);
                }
            }

            System.out.println("============================================================");

        } catch (FileNotFoundException e) {
            System.out.println("ERROR: File not found. Please check the file path.");
        } catch (IOException e) {
            System.out.println("ERROR: Problem reading the file. " + e.getMessage());
        }

        scanner.close(); // Close scanner when done
    }

    // ============================================================
    // METHOD: countOccurrences
    // PURPOSE: Counts how many times 'keyword' appears inside 'text'
    // PARAMETERS:
    //   text    - the string to search in
    //   keyword - the string to search for
    // RETURNS: integer count of occurrences
    // ============================================================
    public static int countOccurrences(String text, String keyword) {
        int count = 0;   // Start at zero
        int index = 0;   // Start searching from position 0

        // Keep searching until indexOf returns -1 (no more matches)
        while ((index = text.indexOf(keyword, index)) != -1) {
            count++;                   // Found one! Add to count
            index += keyword.length(); // Move past this match to find the next
        }

        return count; // Return the total count
    }
}